implement push serivce

caoyue opened the pull request:
## implement push notification serivce
- APNs module
  based on https://github.com/inaka/apns4erl/tree/75f8b8999f5de79b314f61658f2f6d3370399e91
- Certificate file
  - generate certificate file use this [script](https://github.com/inaka/apns4erl/blob/master/priv/test_certs)
  - modify `.pem` certificate file path in `vars.config`
  - a test certificate can be downloaded from [google drive](https://drive.google.com/a/aftech.sg/file/d/0B_FGzvZM_CxVS0I2Zy1zeXg1alE/view?usp=sharing)
- Get feedback from APNs
  - [APNs feedback channel document](https://developer.apple.com/library/ios/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/Chapters/CommunicatingWIthAPS.html#//apple_ref/doc/uid/TP40008194-CH101-SW3)
  - reduce unnecessary message overhead and improve overall system performance
  - connect to feedback channel and try to get deleted tokens everytime pushing messages to APNs
  - "deleted tokens" here means that  intended app does not exist on the device

caoyue committed:
implement push serivce: add/remove push token

caoyue committed:
implement iOS push notification

caoyue committed:
implement feedback handler

caoyue committed:
write config to app.config file

caoyue committed:
add message content to push notification

caoyue committed:
add unique index to apns token

ZekeLu commented:
please merge the `dev` branch so that `Travis CI` will run automatically.

ZekeLu commented on apps/ejabberd/src/mod_push_service.erl:
```diff
@@ -0,0 +1,225 @@
+send_notification(#push_token{type = PushType, token = DeviceToken}, Content) ->
+    case PushType of
+        <<"1">> -> %% iOS push notification
+            mod_push_service_ios:send(binary_to_list(DeviceToken), Content);
```
`binary_to_list`?

caoyue replied:
@ZekeLu because ```apns:send``` function only accept  ```string()``` type device token

ZekeLu commented on apps/ejabberd/src/mod_push_service.erl:
```diff
@@ -0,0 +1,225 @@
+        text ->
+            case Content of
+                undefined ->
+                    <<GroupTitle/binary, NickName/binary, " send a message to you">>;
```
multiple languages support

ZekeLu commented on apps/ejabberd/src/mod_push_service.erl:
```diff
@@ -0,0 +1,225 @@
+            SubContent = binary:part(Content, {0, 100}),
+            <<SubContent/binary, "...">>;
+        false ->
+            binary:part(Content, {0, Size})
```
just return `Content`?

ZekeLu commented on apps/ejabberd/src/mod_push_service_ios.erl:
```diff
@@ -0,0 +1,60 @@
+-module(mod_push_service_ios).
```
Module layout: https://docs.google.com/a/afusion.com/document/d/1ZAoyAElrA8f8nDhf_VFQFDqbOHRRG4JaqGqHix7lMc0/edit

Example: https://github.com/ZekeLu/MongooseIM/blob/master/apps/ejabberd/src/mod_register_aft.erl

ZekeLu commented on apps/ejabberd/src/mod_push_service_ios.erl:
```diff
@@ -0,0 +1,60 @@
+%% Functions
+%% ===========================================
+
+-spec create_notification(DeviceToken :: string(),
```
Can we use `binary` for `DeviceToken`?

ZekeLu commented on apps/ejabberd/src/odbc_push_service.erl:
```diff
@@ -0,0 +1,74 @@
+-include("mod_push_service.hrl").
+
+add(LServer, TokenRecord) ->
+    #push_token{jid = UserJid, token = Token, type = PushType} = TokenRecord,
```
`add(LServer, #push_token{jid = UserJid, token = Token, type = PushType})`

ZekeLu commented on apps/ejabberd/src/odbc_push_service.erl:
```diff
@@ -0,0 +1,74 @@
+add(LServer, TokenRecord) ->
+    #push_token{jid = UserJid, token = Token, type = PushType} = TokenRecord,
+    EJid = ejabberd_odbc:escape(UserJid),
+    IType = integer_to_list(PushType),
```
`integer_to_binary`?

ZekeLu commented on rebar.config:
```diff
@@ -28,7 +28,8 @@
 
-  {pa, ".*", {git, "git://github.com/lavrin/pa.git", "6fc8ef5742"}}
+  {pa, ".*", {git, "git://github.com/lavrin/pa.git", "6fc8ef5742"}},
+  {apns, ".*", {git, "git@github.com:inaka/apns4erl.git", "925dc79c9c0fbce5ca5d2bc5d54bcf49dc1e757a"}}
```
`git@github.com:inaka/apns4erl.git` -> `git@github.com/inaka/apns4erl.git`?

ZekeLu commented on rel/files/app.config:
```diff
@@ -1,11 +1,23 @@
+        ]}
+    ]},
+    %% config for application apns
+    {apns, [
```
use overlay parameters?

ZekeLu commented on apps/ejabberd/src/mod_push_service.erl:
```diff
@@ -0,0 +1,225 @@
+            binary:part(Content, {0, Size})
+    end.
+
+get_host_server() ->
```
do we need this?

caoyue committed:
Add i18n translation of push message

1. Add i18n translation of push message(only chinese now)
2. use overlay template for app.config
3. some code style improve

caoyue committed:
fix bugs in translate module

1. always return language set in config
2. unicode bug

caoyue committed:
remove apns4erl from deps and refactor APNs module in ejabberd application

caoyue committed:
replace jiffy in deps with mochijson2

caoyue committed:
fix error when message have no groupChat attribute

caoyue committed:
return error when received empty token

caoyue committed:
link apns supervisor to ejabberd supervisor

ZekeLu commented on apps/ejabberd/include/mod_apns.hrl:
```diff
@@ -0,0 +1,43 @@
+-record(apns_msg, {
+    id = mod_apns:message_id() :: binary(),
+    expiry = mod_apns:expiry(86400) :: non_neg_integer(), %% default = 1 day
+    device_token :: string(),
```
`binary()`?

ZekeLu commented on apps/ejabberd/priv/mysql.sql:
```diff
@@ -343,3 +343,13 @@ CREATE TABLE privatemode (
+    id int PRIMARY KEY NOT NULL auto_increment,
+    jid varchar(250) CHARACTER SET binary NOT NULL,
+    token varchar(250) CHARACTER SET binary NOT NULL,
+    push_type tinyint unsigned NOT NULL,
```
document `push_type`

ZekeLu commented on apps/ejabberd/priv/zh.msg:
```diff
@@ -398,3 +398,9 @@
 {"Your Jabber account was successfully created.","你的 Jabber 帐户已成功创建."}.
 {"Your Jabber account was successfully deleted.","你的 Jabber 帐户已成功删除."}.
 {"Your messages to ~s are being blocked. To unblock them, visit ~s","您发送给 ~s 的消息已被阻止. 要解除阻止, 请访问 ~s"}.
+{" send a message to you","发来一条消息"}.
```
should use full sentence, such as `~s sends a message to you`

ZekeLu commented on apps/ejabberd/src/ejabberd_sup.erl:
```diff
@@ -37,157 +37,166 @@ start_link() ->
 init([]) ->
     Randoms =
         {randoms,
```
please mind the indent. which format tool do you use?

caoyue committed:
Update APNs module

1. merge newest commit from apns4erl which add apns resend queue feature
2. add some document
3. fix apns feedback handler

ZekeLu committed:
support variables in translation messages
